src/extension.ts

```diff
@@ -57,13 +101,39 @@ async function fetchPackageMetadata(requirement: ProjectNameRequirement): Promis
     return await metadataCache.get(requirement.name)!()
 }
 
+let lastLine = -1
+let lastHover: vscode.Hover | null = null
+let lastLineText = ''
+
 class PyPIHoverProvider implements vscode.HoverProvider {
     async provideHover(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.Hover | null> {
-        const requirement = parsePipRequirementsLineLoosely(document.lineAt(position.line).text)
+        let lineText = document.lineAt(position.line).text
+        if (lastLine === position.line && lineText === lastLineText) {
+            return lastHover
+        }
+
+        const isPyProjectToml =
+            document.languageId === 'toml' && document.fileName.toLowerCase().endsWith('pyproject.toml')
+
+        if (isPyProjectToml) {
```

Comment:
Let's split out the pyproject.toml and pip-requirements branches into separate hover provider methods, e.g. `extractPyProjectDependency` and `extractPipRequirementsDependency` returning  `ProjectNameRequirement | null`. Currently we're running `parsePipRequirementsLineLoosely` even for pyproject.toml files